### DISCUSSION TITLE
Prefetching infinite query triggers infinite re-renders

### DISCUSSION THREAD

**Discussion by aniketbiswas21** [TanStack/query#3866] [Q&A] [answered]
### Describe the bug

I was trying to implement an infinite scroller using `useInfiniteQuery`. To improve the user experience, I decided to prefetch the next cursor's data using `prefetchInfiniteQuery`, whenever the response from the first one is returned

But, prefetching infinite query leads to infinite rendering. Is there a way to achieve the above?

### Your minimal, reproducible example

https://codesandbox.io/s/black-hill-4g2n7y?file=/src/List.js

### Steps to reproduce

1. Try uncommenting the onSuccess callback in the codesandbox. Would lead to infinite rendering

### Expected behavior

The next route is prefetched and when the user clicks on the loadMore button, the data is returned from cache.

### How often does this bug happen?

Every time

### Platform

- OS: macOS v12.3.1

### react-query version

4.0.0

**Comment by TkDodo** [comment, marked as answer]
without an exit condition, this will of course trigger an infinite render, because every fetch triggers onSuccess, which triggers another fetch, which triggers onSuccess, ...

the question is _when_ you do you want to fetch the next page? There is no prefetching once a query is displayed because fetching the next page will just display it as well.

what we do in the examples is use an intersection observer and fetch the next page when this element gets into the viewport. This can be at the very end of the list, or e.g. 5 entries before that. That way, when the user scrolls to the "almost bottom", the fetch will happen, and by the time they are at the bottom, they will see the new entries.

**Reply by aniketbiswas21** [reply]
@TkDodo So, let's say after fetching the first page, if I want to fetch the next page in background what would be the ideal way to deal with this? Considering onSuccess is triggered for prefetching too?

P.S. The user is yet to click on the loadMore button

**Reply by TkDodo** [reply]
what would "in the background" mean here? infinite queries only have one cache that consists of multiple pages; as soon as you fetch the second page, it will be visible on the screen.

maybe you are looking for a normal useQuery like here:
https://github.com/TanStack/query/blob/9537fe961c786604f09ca3ed8d24d93eeb7eedf3/examples/react/pagination/pages/index.js#L36-L43

**Reply by aniketbiswas21** [reply]
Ok, thanks.

But, I still cannot understand if we cannot prefetch the next page, for which case `prefetchInfiniteQuery` is used?

**Reply by TkDodo** [reply]
to prefetch it _before_ you render it.

**Reply by aniketbiswas21** [reply]
ok understood thanks 👍
